File: part1/src/main/java/com/peterlaurence/book/javatokotlin/callbackpattern/fragments/PurchaseViewModelFactory.java
package com.peterlaurence.book.javatokotlin.callbackpattern.fragments;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.peterlaurence.book.javatokotlin.callbackpattern.core.java.PurchasesProviderImpl;
import com.peterlaurence.book.javatokotlin.callbackpattern.fragments.PurchasesViewModel.PurchasesProvider;
import com.peterlaurence.book.javatokotlin.callbackpattern.fragments.PurchasesViewModel.BillingClient;

import com.peterlaurence.book.javatokotlin.callbackpattern.core.java.BillingClientImpl;

public class PurchaseViewModelFactory implements ViewModelProvider.Factory {
    private PurchasesProvider provider = new PurchasesProviderImpl();
    private BillingClient billingClient = new BillingClientImpl(provider);

    private final String user = "user"; // Get in from registration service

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(PurchasesViewModel.class)) {
            return (T) new PurchasesViewModel(billingClient, user);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: booksamples/src/main/java/com/peterlaurence/book/pawk/callbackpattern/fragments/PurchaseViewModelFactory.kt
package com.peterlaurence.book.pawk.callbackpattern.fragments

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.peterlaurence.book.pawk.callbackpattern.core.BillingClientImpl
import com.peterlaurence.book.pawk.callbackpattern.core.PurchasesProviderImpl
import com.peterlaurence.book.pawk.callbackpattern.fragments.PurchasesViewModel.BillingClient
import com.peterlaurence.book.pawk.callbackpattern.fragments.PurchasesViewModel.PurchasesProvider

class PurchaseViewModelFactory : ViewModelProvider.Factory {
    private val provider: PurchasesProvider = PurchasesProviderImpl()
    private val billingClient: BillingClient = BillingClientImpl(provider)
    private val user = "user" // Get in from registration service

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(PurchasesViewModel::class.java)) {
            return PurchasesViewModel(billingClient, user) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}